### PR TITLE
port OSD from mpc-be

### PR DESCRIPTION
**Pull request by adipose**
This solves some issues with the existing OSD.  I kept the existing one in the source tree, but I didn't really port any behavior from the old one to the new one, unless it was required to keep existing functionality (showtime).  

It behaves slightly differently, and I am not 100% sure if any features in recent years might be lost.  But VMROSD hasn't changed much, I don't think.

The main thing is, it looks much better in windowed mode during resizes (MPCVR and EVRCP at least).  MADVR is using a different OSD.

### CODE REVIEW

**Comment by clsid2** [comment]
This would work for audio only files as well, right?
(if "video" area is large enough)

**Comment by adipose** [comment]
> This would work for audio only files as well, right? (if "video" area is large enough)

Yes, I think so, I see code to allow it.

**Comment by clsid2** [comment]
Then a few audioonly checks may need to be removed.

Plus a little size sanity check before trying to render OSD. Because with empty logo the minimum size is 16x16, which would be too small.

Might also need a test with MPCVR with exclusive mode enabled.

Another related possible future feature addition would be to use the D3D exclusive mode seekbar from VMR/EVR also for MPCVR when it is using exclusive mode. MPCBE has some code to handle exclusive mode for MPCVR.

**Comment by adipose** [comment]
Oh, I see what you mean.

I did test MPCVR exclusive.  However, I observed that MPC-BE has some code to control how MPCVR exclusive is turned on and off, which may not align with the D3D setting.

**Comment by adipose** [comment]
Here's what it does when the window is too small.

image

Because it's a child window, any overlap will just be clipped (but it seems to preemptively crop it).

**Comment by adipose** [comment]
For audio files, it seemed to work ok:

image

**Comment by clsid2** [comment]
> Here's what it does when the window is too small.
> 
> image
> 
> Because it's a child window, any overlap will just be clipped (but it seems to preemptively crop it).

Partial OSD is a bit ugly. I would prefer to just hide it when area is small. Because it is also user intention to have small area and not show anything. What would be a sane minimum height for showing OSD? Maybe 80px

**Comment by adipose** [comment]
> > Here's what it does when the window is too small.
> > image
> > Because it's a child window, any overlap will just be clipped (but it seems to preemptively crop it).
> 
> Partial OSD is a bit ugly. I would prefer to just hide it when area is small. Because it is also user intention to have small area and not show anything. What would be a sane minimum height for showing OSD? Maybe 80px

As the OSD doesn't resize, we could just refuse to show it if smaller than that size.  MADVR OSD is a bit shorter.

File: src/DSUtil/DSUtil.cpp [hunk 1]
```diff
@@ -1834,6 +1834,18 @@ CString ReftimeToString3(const REFERENCE_TIME& rtVal)
     return strTemp;
 }
 
+//for compatibility with mpc-be ReftimeToString2, which has option to exclude hours
```
Comment (clsid2) [inline src/DSUtil/DSUtil.cpp:1837]: This function does not seem to be used?

Comment (clsid2) [reply src/DSUtil/DSUtil.cpp:1837]: OSD.cpp/.h missing in patch

**Comment by clsid2** [comment]
Will look at this later. I am a bit busy this month.

**Comment by clsid2** [comment]
PR contains unwanted LAV changes.

**Comment by adipose** [comment]
> PR contains unwanted LAV changes.

Hopefully fixed now.

**Comment by clsid2** [comment]
I am not sure if I really like the transparency. It can make OSD more difficult to read, and text quality seems a bit worse due to it. Old OSD style with solid background and border looks better.

osd_comparison

EVR-CP D3D exclusive fullscreen seekbar is also not shown anymore.

**Comment by adipose** [comment]
In MPCBE transparency is 100/255 by default but can be configured.

I kept the code but I didn't expose the setting. Default could be 0 instead.

**Comment by clsid2** [comment]
Yes, change to 0 and add border. Blending/drawing code could maybe also be optimized when no transparency is used.

No need for an option. Maybe in future.

**Comment by clsid2** [comment]
I don't see a border yet?
It also looks like the text is thicker in dark theme vs light theme.

**Comment by adipose** [comment]
> I don't see a border yet? It also looks like the text is thicker in dark theme vs light theme.

I'm still looking at the OSD.  Working on the D3D problems, which are multiple (some fixed now).  In my screenshot, however, I had a border in the OSD.

image

**Comment by MadokaAyukawa** [comment]
The new OSD flickers when I skip forwards or backwards with the time displayed. It also isn't displayed until 3-4 seconds have passed when starting a new video in full screen (double-clicking on a video to auto-open HC in full screen). Also, would you consider making the border optional as I tend to have the time on in movies where the OSD/Time is in the black bar, and the border makes it stand out a lot more.

**Comment by clsid2** [comment]
I don't get any flickering when seeking.

When launching in fullscreen, there is an issue related to filename display, which is normally shown first 3 seconds. I see OSD appear for a fraction of a seconds when launching. It is more obvious in light theme and dark video.

When closing player, the OSD also stays on screen too long. Should be hidden sooner.

**Comment by adipose** [comment]
Border is optional but the setting is not currently exposed.

The flicker when seeking I am not sure about.  Unless somehow the window is hidden and shown, it shouldn't flicker, as it uses a memdc.

Can you record the flickering?

**Comment by MadokaAyukawa** [comment]
I'm sorry, I don't know how to make screen recordings.

**Comment by adipose** [comment]
> I'm sorry, I don't know how to make screen recordings.

snagit

**Comment by adipose** [comment]
> I'm sorry, I don't know how to make screen recordings.

Snagit

**Comment by bibsp** [comment]
@adipose `hide OSD when space unavailable` this patch is creating problem.
If there is Horizontal space for OSD but the file name/title is long, then the complete OSD is hidden.
It should only hide the OSD if there is no Vertical space.

> Here's what it does when the window is too small.
> 
> image

For this situation only.
Currently it does not show file name/chapter name.

**Comment by adipose** [comment]
> @adipose `hide OSD when space unavailable` this patch is creating problem. If there is Horizontal space for OSD but the file name/title is long, then the complete OSD is hidden. It should only hide the OSD if there is no Vertical space.
> 
> > Here's what it does when the window is too small.
> 
> For this situation only. Currently it does not show file name/chapter name.

If @clsid2 agrees you can submit a patch to ignore the lack of horizontal space.

**Comment by bibsp** [comment]
I don't know why anyone will disagree to this. All softwares (Not only mpc but image viewer, file manager etc) follow the same logic, if name is long, show partial name.
You can't cut a letter vertically, but you can show partial name horizontally, this is logical. Don't get angry, think about it. Even old OSD was like that.

And secondly I'm not coder. I have just noticed it, so mentioned here. Decision is yours.

**Comment by adipose** [comment]
> And secondly I'm not coder. I have just noticed it, so mentioned here. Decision is yours.

You identified the right commit.  I bet you could figure it out!

**Comment by clsid2** [comment]
Showing partial OSD horizontally is fine. Player has a minimum width anyway.

**Comment by adipose** [comment]
https://github.com/clsid2/mpc-hc/pull/2783

**Comment by clsid2** [comment]
Are you able to reproduce the issue where OSD disappears later than the window when closing?

When OSD is shown and you switch to a background window, then it also stays visible sometimes. Easiest to reproduce with the filename display directly after opening.

**Comment by adipose** [comment]
> Are you able to reproduce the issue where OSD disappears later than the window when closing?

No.  It disappears right away.

I can believe it though, since it's a separate window.  Maybe I can artificially make it happen.

**Comment by adipose** [comment]
```
    m_OSD.Stop();
    ShowWindow(SW_HIDE);
```

If you add this line to `CMainFrame::OnClose()`, does it help any?

Or possibly:

```
    m_wndView.ShowWindow(SW_HIDE);
    ShowWindow(SW_HIDE);
```

**Comment by clsid2** [comment]
`m_OSD.Stop()` did not work until I changed
`COSD::ClearMessage` to call `ShowWindow(SW_HIDE);` directly instead of indirectly through `PostMessageW(WM_OSD_HIDE);`
Seems like the window message does not get processed anymore.
Any reason why window message was used?

**Comment by adipose** [comment]
I don't know why, but a typical reason would be if the parent call was already happening in a recursive loop.

**Comment by clsid2** [comment]
Just calling `m_OSD.OnHide()` in `CMainFrame::OnClose()` works too.

**Comment by clsid2** [comment]
> When OSD is shown and you switch to a background window, then it also stays visible sometimes. Easiest to reproduce with the filename display directly after opening.

And when restoring from minimized window, the OSD is shown quicker than the restore animation takes to complete. Quite funny too seen the main window expand upwards towards the OSD when animation is a bit slow. This one less important than the other issue. MPC-BE also has the restore issue, but not the above issue.

**Comment by adipose** [comment]
I considered it might be during animation.  But even with it on, I couldn't reproduce it.

**Comment by adipose** [comment]
I reproduced it on windows 7 only.

**Comment by adipose** [comment]
Try this and let me know.  It seemed to solve the problem.

```
BOOL COSD::OnEraseBkgnd(CDC* pDC)
{
    return FALSE;
}
```

**Comment by clsid2** [comment]
Does not solve it.

Weird thing also is that it mostly only happens with the filename display at opening. Volume OSD seems fine, unless I trigger it directly at start. When reloading file the filename display is also fine most of the time. So maybe something different with OSD window state at initial launch?

During testing I also got several times an issue that player window would not show after switching to other window and back. Pressing Alt (to activate menu bar) makes it re-appear.

**Comment by adipose** [comment]
Was your testing all windows 7?

I used a 7 VM and reproduced it.  But changing the return value of erasebg worked.

I am surprised it didn't help because after changing that I could not repro it.  It's a transparent layered window but maybe it's not moving with the parent.

**Comment by adipose** [comment]
```
if (SysVersion::IsWin8orLater()) {
        dwStyle   |= WS_CHILD;
```

Maybe this is an issue with Win7.

**Comment by adipose** [comment]
I will test later today but if the behavior is due to the win7 specific code then maybe I can reproduce it on win10 by forcing it on.

**Comment by clsid2** [comment]
Yes, Win7.

MPC-BE 1.7.1 does not have the issue here.

**Comment by adipose** [comment]
And using mpcvr?

**Comment by adipose** [comment]
If you have time can you see if any older MPCBE had the issue?

**Comment by clsid2** [comment]
Happens with MPCVR, EVR-CP, and even VMR9.

MPCBE 1.6.11 also ok.

**Comment by adipose** [comment]
Hmm ok.  I will see if I can repro in Windows 10.

**Comment by adipose** [comment]
https://github.com/clsid2/mpc-hc/pull/2787

I still need to determine why this doesn't affect BE.  But I was able to reproduce the problem in Windows 10 by forcing the Windows 7 style window and introducing an artificial delay.  What this does is hide the OSD when the player is minimized or not visible.  It only applies when the Win7 behavior is active.

**Comment by adipose** [comment]
I can sort of repro it in mpcbe using the same hack.  Not quite the same because it doesn't display things as long.

**Pull request merged**